Add roundOutwards to Fixed18.div, roundUp to UFixed18.div

0xfennel opened the pull request:
In signed arithmetic, we use `roundOutwards` instead of `roundUp` because "up" is ambiguous when negative numbers come into play.

kbrizzle commented on contracts/number/types/UFixed18.sol:
```diff
@@ -104,6 +104,20 @@ library UFixed18Lib {
+     * @param roundUp Whether to round result up to the next integer
+     * @return Resulting divided unsigned fixed-decimal
+     */
+    function div(UFixed18 a, UFixed18 b, bool roundUp) internal pure returns (UFixed18) {
```
since we're kind of doing the overload backwards here, what if instead of the flag we just have two functions: `div` and `divOut`?

0xfennel replied:
Sure, I assume same goes for `Fixed18` as well?

kbrizzle commented on contracts/mocks/MockFixed18.sol:
```diff
@@ -60,6 +60,10 @@ contract MockFixed18 {
         return Fixed18Lib.div(a, b);
     }
 
+    function div(Fixed18 a, Fixed18 b, bool roundOutwards) external pure returns (Fixed18) {
```
note: we need to do this for all of the `.mul()` functions as well since they can have the same problem while dividing by the `BASE`

0xfennel replied:
`.mul()` and the two `.mulDiv()`s, right?

kbrizzle replied:
oh yes those too!

kbrizzle commented on test/unit/number/Fixed18.test.ts:
```diff
@@ -135,29 +135,79 @@ describe('Fixed18', () => {
+
+    it('divs and rounds away from 0', async () => {
+      expect(await fixed18['div(int256,int256,bool)'](21, utils.parseEther('10'), true)).to.equal(3)
+      expect(await fixed18['div(int256,int256,bool)'](-21, utils.parseEther('10'), true)).to.equal(-3)
```
can you test negative divisor here too

kbrizzle commented on contracts/number/types/Fixed18.sol:
```diff
@@ -120,6 +120,22 @@ library Fixed18Lib {
+
+        int256 numerator = Fixed18.unwrap(a) * BASE;
+        int256 hasRemainder = (numerator % Fixed18.unwrap(b) != 0) ? int256(1) : int256(0);
+        return Fixed18.wrap(numerator / Fixed18.unwrap(b) + hasRemainder * sign(a) * sign(b));
```
can we simplify this section to something like:

`sign(a.mul(b)).mul(a.abs().divCeil(b.abs()))`

I'm not sure if this would add a ton of gas though?

0xfennel replied:
Couple of comments here:
- We have to add in a zero-denominator check since OpenZeppelin's `ceilDiv` doesn't revert if `a` and `b` are both 0.
- `sign(a) * sign(b)` is a little safer in case `a.mul(b) overflows`. But the latter saves 41 gas (maybe because `sign` is expensive, haven't looked into it yet). @kbrizzle will defer to you on which one we go with here.
- Surprisingly, even the former saves 65 gas! (23025 vs 22960 gas with viaIR turned on). So I'll switch to something like:
```solidity
function divOut(Fixed18 a, Fixed18 b) internal pure returns (Fixed18) {
    if (Fixed18.unwrap(b) == 0) {
        revert Fixed18DivideByZeroError();
    }

    uint256 unsignedQuotient = Math.ceilDiv(UFixed18.unwrap(a.abs()) * uint256(BASE), UFixed18.unwrap((b.abs())));
    return Fixed18.wrap(sign(a) * sign(b) * (int256)(unsignedQuotient));
}
```

kbrizzle replied:
I see yeah the naive `a * b` restricts the range more than it needs to. How about something like this, I think this is super clean. Basically trying to minimize the number of place we have custom math under the hood, and reuse other stuff as long as it doesn't cause tons of extra gas.

```
function divOut(Fixed18 a, Fixed18 b) internal pure returns (Fixed18) {
    if (Fixed18.unwrap(b) == 0) revert Fixed18DivideByZeroError();

    return Fixed18.from(sign(a) * sign(b), a.abs().divOut(b.abs()));
}
```

0xfennel replied:
Nice, using `from` is an awesome idea!

0xfennel replied:
Since this uses `UFixed18`'s `divOut` which does divide-by-zero reverts correctly, we also won't need the zero-denominator check :)

kbrizzle replied:
yeah exactly!

0xfennel commented:
Ok, addressed all the comments. Added `*Out()` functions for `mul`, `div`, `unsafeDiv`, and `mulDiv`. Abstracted some of the utility functions into a separate `Math` library (need a second opinion on the name and organization...).

kbrizzle commented on contracts/number/Math.sol:
```diff
@@ -0,0 +1,45 @@
+     * @return Resulting quotient
+     */
+    function divOut(uint256 a, uint256 b) internal pure returns (uint256) {
+        if (a == 0 && b != 0) return 0;
```
something really tricky we could do here would be to actually throw a real divide-by-zero error when we see `b==0` using this: https://docs.soliditylang.org/en/v0.8.17/control-structures.html#panic-via-assert-and-error-via-require

Then we can just use `Math.ceilDiv` directly from OZ:

```
error Panic(uint256 code);

function divOut(uint256 a, uint256 b) internal pure returns (uint256) {
    if (b == 0) Panic(0x12);
    return Math.ceilDiv(a, b);
}
```

kbrizzle replied:
not 100% sure if this works, but if it does I think this is cleaner

0xfennel replied:
This doesn't work. If you like using `Math.ceilDiv` directly, though, then we can just revert with a custom error instead.

0xfennel replied:
Turns out, using `Math.ceilDiv` actually uses less gas somehow, so we decided to do this.

kbrizzle commented on contracts/number/Math.sol:
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: Apache-2.0
```
Let's name this `NumberUtils.sol` we do have some precedent with that scheme elsewhere in the `root` codebase - and then we don't have to overload the naming when we import.

kbrizzle commented on contracts/number/types/UFixed18.sol:
```diff
@@ -94,6 +95,16 @@ library UFixed18Lib {
+     * @return Resulting multiplied unsigned fixed-decimal
+     */
+    function mulOut(UFixed18 a, UFixed18 b) internal pure returns (UFixed18) {
+        return UFixed18.wrap(Math.ceilDiv(UFixed18.unwrap(a) * UFixed18.unwrap(b), BASE));
```
use our own Math.divOut here

kbrizzle commented on contracts/number/types/UFixed18.sol:
```diff
@@ -104,6 +115,17 @@ library UFixed18Lib {
+     * @return Resulting divided unsigned fixed-decimal
+     */
+    function divOut(UFixed18 a, UFixed18 b) internal pure returns (UFixed18) {
+        if (isZero(a) && !isZero(b)) return ZERO;
```
you can rely on the internal check here

kbrizzle commented on contracts/number/types/UFixed18.sol:
```diff
@@ -141,6 +190,17 @@ library UFixed18Lib {
+     * @return Resulting computation
+     */
+    function muldivOut(UFixed18 a, UFixed18 b, UFixed18 c) internal pure returns (UFixed18) {
+        return UFixed18.wrap(Math.ceilDiv(UFixed18.unwrap(a) * UFixed18.unwrap(b), UFixed18.unwrap(c)));
```
use our own Math.divOut here
